### PR TITLE
Preemptible slow path

### PR DESCRIPTION
**Pull request by jvilk-stripe**
This is an "epic" of a PR that has accumulated changes from several PRs to make landing easier.

The changes include:
* Move cancellation from preprocessor thread to message processing thread.
* Change preprocessor to operate on _tasks_ rather than _messages_.
* Change the `LSPTask` interface to implement nearly all of the behavior of each LSP method
  * e.g., it now dictates behavior during preprocessing and indexing, and if it even needs to interface with the typechecker
* Schedules a LSPQueuePreemptionTask during the slow path that runs all of the tasks that can preempt that are at the top of the scheduling queue.
* Cancels a LSPQueuePreemptionTask if the task queue mutates to a state where the first task can no longer preempt (e.g., a slow path edit that would cancel).

The end result is a cleaner codebase with support for preemption without degrading the performance of cancelation.

### Motivation

Reduces the amount of time Sorbet's language server spends in a state where it is unresponsive to user requests.

### Test plan

See included automated tests.

### CODE REVIEW

File: core/lsp/TypecheckEpochManager.cc [hunk 1]
```diff
@@ -13,22 +13,16 @@ void TypecheckEpochManager::assertConsistentThread(optional<thread::id> &expecte
     }
 }
 
-void TypecheckEpochManager::startCommitEpoch(u4 fromEpoch, u4 toEpoch) {
+void TypecheckEpochManager::startCommitEpoch(u4 epoch) {
```
Comment (jvilk-stripe) [inline core/lsp/TypecheckEpochManager.cc]: Before this change, the preprocessor needed to know the full range of edit epochs contained in the currently typechecking slow path (for reasons related to `TimeTravelingGlobalState`). It was pretty hairy, and confused the definition of 'epoch'.

We now only need to know the epoch of the latest slow path, which corresponds to the epoch of the latest edit in the combined file updates.

File: core/lsp/TypecheckEpochManager.cc [hunk 2]
```diff
@@ -52,7 +46,8 @@ TypecheckEpochManager::TypecheckingStatus TypecheckEpochManager::getStatus() con
 }
 
 bool TypecheckEpochManager::tryCancelSlowPath(u4 newEpoch) {
-    assertConsistentThread(preprocessThreadId, "TypecheckEpochManager::tryCancelSlowPath", "preprocessThread");
+    assertConsistentThread(messageProcessingThreadId, "TypecheckEpochManager::tryCancelSlowPath",
```
Comment (jvilk-stripe) [inline core/lsp/TypecheckEpochManager.cc]: We moved cancelation to the message processing thread.

File: main/lsp/LSPPreprocessor.cc [hunk 3]
```diff
@@ -125,64 +89,14 @@ void LSPPreprocessor::mergeFileChanges(absl::Mutex &mtx, QueueState &state) {
         }
     }
     ENFORCE(pendingRequests.size() + requestsMergedCounter == originalSize);
-
-    // Check if we should cancel the slow path.
```
Comment (jvilk-stripe) [inline main/lsp/LSPPreprocessor.cc]: Goodbye, hairy logic! The preprocessor no longer needs to think about time travel or cancelation.

File: main/lsp/LSPPreprocessor.cc [hunk 4]
```diff
@@ -125,64 +89,14 @@ void LSPPreprocessor::mergeFileChanges(absl::Mutex &mtx, QueueState &state) {
...
-    return ttgs.getGlobalState().deepCopy();
 }
 
 unique_ptr<LSPMessage> LSPPreprocessor::makeAndCommitWorkspaceEdit(unique_ptr<SorbetWorkspaceEditParams> params,
                                                                    unique_ptr<LSPMessage> oldMsg) {
-    ttgs.commitEdits(params->updates);
-    if (!params->updates.canTakeFastPath) {
-        params->updates.updatedGS = getTypecheckingGS();
+    for (auto &u : params->updates) {
+        fileContents[u->path()] = u;
```
Comment (jvilk-stripe) [inline main/lsp/LSPPreprocessor.cc]: It does need to track the latest file contents so we can process diffs sent from clients. We're tracking `shared_ptr<core::File>`, which are immutable objects shared with `GlobalState`, so this is _not_ adding O(workspace size) memory overhead. It's adding O(size_of_pointer * # of files in workspace).

File: main/lsp/LSPPreprocessor.cc [hunk 5]
```diff
@@ -197,7 +111,6 @@ bool LSPPreprocessor::ensureInitialized(LSPMethod method, const LSPMessage &msg)
         return true;
     }
     config->logger->error("Serving request before got an Initialize & Initialized handshake from IDE");
-    vector<unique_ptr<LSPMessage>> responses;
```
Comment (jvilk-stripe) [inline main/lsp/LSPPreprocessor.cc]: This variable was unused 🤷‍♂

File: main/lsp/LSPPreprocessor.cc [hunk 6]
```diff
@@ -266,16 +177,7 @@ void LSPPreprocessor::preprocessAndEnqueue(QueueState &state, unique_ptr<LSPMess
             break;
         }
         case LSPMethod::Initialized: {
-            InitializedParams &params = *get<unique_ptr<InitializedParams>>(msg->asNotification().params);
```
Comment (jvilk-stripe) [inline main/lsp/LSPPreprocessor.cc]: We now perform indexing and file hashing via the `IndexAndHashWorkspaceTask` run on the typechecker thread. This is the blocking part of `initialization` that we talked about before.

File: main/lsp/LSPPreprocessor.cc [hunk 7]
```diff
@@ -345,112 +243,68 @@ void LSPPreprocessor::preprocessAndEnqueue(QueueState &state, unique_ptr<LSPMess
...
         if (!config->isFileIgnored(localPath) && !openFiles.contains(localPath)) {
-            updates.updatedFiles.push_back(make_shared<core::File>(
-                move(localPath), readFile(localPath, *config->opts.fs), core::File::Type::Normal, v));
+            edit->updates.push_back(make_shared<core::File>(move(localPath), readFile(localPath, *config->opts.fs),
+                                                            core::File::Type::Normal, v));
         }
     }
-}
-
-void LSPPreprocessor::mergeEdits(LSPFileUpdates &to, LSPFileUpdates &from) {
```
Comment (jvilk-stripe) [inline main/lsp/LSPPreprocessor.cc]: The preprocessor no longer deals with `LSPFileUpdates` objects, which contain a lot of stuff the preprocessor no longer cares about (like parsed ASTs). It only deals with LSP messages, and merging adjacent `SorbetWorkspaceEdit` messages.

File: main/lsp/LSPTypechecker.cc [hunk 8]
```diff
@@ -43,6 +43,8 @@ LSPTypechecker::LSPTypechecker(std::shared_ptr<const LSPConfiguration> config)
 
 void LSPTypechecker::initialize(LSPFileUpdates updates, WorkerPool &workers) {
     ENFORCE(this_thread::get_id() == typecheckerThreadId, "Typechecker can only be used from the typechecker thread.");
+    ENFORCE(!this->initialized);
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.cc]: I beefed up some enforces to ensure that we never try to initialize twice, and that we never try to use an uninitialized global state.

File: main/lsp/LSPTypechecker.cc [hunk 9]
```diff
@@ -547,6 +549,62 @@ void LSPTypechecker::changeThread() {
     typecheckerThreadId = newId;
 }
 
+vector<core::FileHash> LSPTypechecker::computeFileHashes(const LSPConfiguration &config,
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.cc]: Static helper method moved from `TimeTravelingGlobalState`. Virtually unchanged.

File: main/lsp/LSPTypechecker.h [hunk 10]
```diff
@@ -148,10 +158,13 @@ class LSPTypechecker final {
  * Provides lambdas with a set of operations that they are allowed to do with the LSPTypechecker.
  */
 class LSPTypecheckerDelegate {
-    WorkerPool &workers;
     LSPTypechecker &typechecker;
 
 public:
+    /** The WorkerPool on which work will be performed. If the task is multithreaded, the pool will contain multiple
+     * worker threads. */
+    WorkerPool &workers;
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypechecker.h]: Some of the new tasks I added need direct access to the WorkerPool to distribute work amongst worker threads. They don't involve `LSPTypechecker`, so it didn't make sense to add a layer of delegation to them.

File: main/lsp/LSPTypecheckerCoordinator.cc [hunk 11]
```diff
@@ -52,41 +52,43 @@ class TypecheckerTask final : public core::lsp::Task {
 
 class InitializeTask : public LSPTask {
     LSPTypechecker &typechecker;
-    WorkerPool &workers;
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypecheckerCoordinator.cc]: Now that `TypecheckerDelegate` lets the task access `workers`,it doesn't need to store it separately.

File: main/lsp/LSPTypecheckerCoordinator.cc [hunk 12]
```diff
@@ -52,41 +52,43 @@ class TypecheckerTask final : public core::lsp::Task {
...
 
-    void run(LSPTypecheckerDelegate &_) override {
-        typechecker.initialize(move(updates), workers);
+    void run(LSPTypecheckerDelegate &tcd) override {
+        typechecker.initialize(move(updates), tcd.workers);
     }
 };
 
-class SlowPathTypecheckTask : public LSPTask {
+class SlowPathTypecheckTask : public core::lsp::Task {
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypecheckerCoordinator.cc]: Now that the slow path is non-blocking, this task doesn't need to use the delegate or common functionality in `LSPTask`, so I made it a regular `Task` . 🎉

File: main/lsp/LSPTypecheckerCoordinator.cc [hunk 13]
```diff
@@ -129,14 +131,18 @@ void LSPTypecheckerCoordinator::syncRun(unique_ptr<LSPTask> task) {
     wrappedTask->blockUntilComplete();
 }
 
-void LSPTypecheckerCoordinator::initialize(unique_ptr<InitializedParams> params) {
-    // TODO: Split into a sync and async task when we land preemptible slow path, where the typecheck is async.
-    syncRun(make_unique<InitializeTask>(*config, typechecker, workers, move(params->updates)));
+void LSPTypecheckerCoordinator::initialize(LSPFileUpdates initialUpdate) {
+    // TODO: Make an async task when we land preemptible slow path, where the typecheck is async.
```
Comment (jvilk-stripe) [inline main/lsp/LSPTypecheckerCoordinator.cc]: It doesn't make sense to make initialization asynchronous now, as the next message would end up blocked anyway.

File: main/lsp/TimeTravelingGlobalState.cc [hunk 14]
```diff
@@ -1,339 +0,0 @@
-#include "main/lsp/TimeTravelingGlobalState.h"
-#include "core/Unfreeze.h"
-#include "main/lsp/LSPPreprocessor.h"
-#include "main/pipeline/pipeline.h"
-
-using namespace std;
-
-namespace sorbet::realmain::lsp {
-TimeTravelingGlobalState::TimeTravelingGlobalState(const shared_ptr<LSPConfiguration> &config,
```
Comment (jvilk-stripe) [inline main/lsp/TimeTravelingGlobalState.cc]: This class was a huge source of complexity. I'm glad it's gone.

File: main/lsp/json_types.cc [hunk 15]
```diff
@@ -247,4 +247,26 @@ LSPFileUpdates LSPFileUpdates::copy() const {
     return copy;
 }
 
+void SorbetWorkspaceEditParams::merge(SorbetWorkspaceEditParams &newerParams) {
```
Comment (jvilk-stripe) [inline main/lsp/json_types.cc]: This is the method the preprocessor calls to merge adjacent edits in the queue.

File: main/lsp/lsp.h [hunk 16]
```diff
@@ -36,16 +36,31 @@ class LSPLoop {
...
 
     void processRequestInternal(LSPMessage &msg);
 
     /** Returns `true` if 5 minutes have elapsed since LSP last sent counters to statsd. */
     bool shouldSendCountersToStatsd(std::chrono::time_point<std::chrono::steady_clock> currentTime) const;
     /** Sends counters to statsd. */
     void sendCountersToStatsd(std::chrono::time_point<std::chrono::steady_clock> currentTime);
-    /** Helper method: If message is an edit taking the slow path, and slow path cancelation is enabled, signal to
-     * GlobalState that we will be starting a commit of the edits. */
-    void maybeStartCommitSlowPathEdit(const LSPMessage &msg) const;
```
Comment (jvilk-stripe) [inline main/lsp/lsp.h]: No longer needed; the `LSPTypecheckerCoordinator` now deals with this as a part of slow path typechecking.

File: main/lsp/lsp.h [hunk 17]
```diff
@@ -36,16 +36,31 @@ class LSPLoop {
...
+
+    /** Commits the given edit to `initialGS`, and returns a canonical LSPFileUpdates object containing indexed trees
+     * and file hashes. Also handles canceling the running slow path. */
+    LSPFileUpdates commitEdit(SorbetWorkspaceEditParams &edit);
```
Comment (jvilk-stripe) [inline main/lsp/lsp.h]: I hate that I overuse the term 'commit'. I don't know a better word here.

File: main/lsp/request_dispatch.cc [hunk 18]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+        prodCategoryCounterAdd("lsp.messages.processed", "sorbet/mergedEdits", updates.editCount - 1);
+    }
+};
+
+class IndexAndHashWorkspaceTask final : public LSPTask {
+    unique_ptr<core::GlobalState> &initialGS;
+    vector<core::FileHash> &globalStateHashes;
+    LSPFileUpdates &updates;
+
+    vector<ast::ParsedFile> indexFromFileSystem(WorkerPool &workers, std::unique_ptr<KeyValueStore> kvstore) {
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: Virtually identical to `TimeTravelingGlobalState::indexFromFileSystem`.

File: main/lsp/request_dispatch.cc [hunk 19]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+        // Temporarily replace error queue, as it asserts that the same thread that created it uses it and we're
+        // going to use it on typechecker thread for this one operation.
+        auto savedErrorQueue = initialGS->errorQueue;
+        initialGS->errorQueue = make_shared<core::ErrorQueue>(savedErrorQueue->logger, savedErrorQueue->tracer);
+        // Enforce that this is only run once.
+        ENFORCE(globalStateHashes.empty());
+
+        vector<ast::ParsedFile> indexed;
+        {
+            Timer timeit(config.logger, "initial_index");
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: This block moved from `LSPPreprocessor.cc`.

File: main/lsp/LSPPreprocessor.cc (hunk 4 above)
Comment (jvilk-stripe) [reply main/lsp/LSPPreprocessor.cc]: Actually, I'm going to push an update; `fileContents` should only contain the contents of open files, not all files. My bad.

File: main/lsp/request_dispatch.cc [hunk 20]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+const core::FileHash &findHash(int id, const vector<core::FileHash> &globalStateHashes,
+                               const UnorderedMap<int, core::FileHash> &overriddingStateHashes) {
+    const auto it = overriddingStateHashes.find(id);
+    if (it == overriddingStateHashes.end()) {
+        return globalStateHashes[id];
+    }
+    return it->second;
+}
+
+bool canTakeFastPath(
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: Mostly unchanged from `TimeTravelingGlobalState`, but introduces the notion of `overriddingStateHashes`.

Three important things to know:
1. These file hashes are what we use to determine if edits can take the fast or slow path.
2. `globalStateHashes` contains the latest file hashes as of this update.
3. If a slow path is running, we need to be able to tell if an edit would take the fast path with the edits typechecking on the slow path.

In order to accomplish 3 given 1 and 2, `overridingStateHashes` contains the file hashes evicted from `globalStateHashes` since the slow path began. It lets us make a fast path decision in the immediate past.

File: main/lsp/request_dispatch.cc [hunk 21]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+                    return false;
+                }
+            }
+        }
+    }
+    logger.debug("Taking fast path");
+    return true;
+}
+
+UnorderedMap<int, core::FileHash> mergeEvictions(const UnorderedMap<int, core::FileHash> &olderEvictions,
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: Merges file hash evictions from two edits. Needed to make fast path decisions in the past, e.g. `olderEvictions` will contain the hashes evicted by the running slow path and `newerEvictions` will contain the hashes evicted by the update we are currently processing.

File: main/lsp/request_dispatch.cc [hunk 22]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+            } else {
+                // This file update adds a new file to GlobalState.
+                update.hasNewFiles = true;
+                fref = initialGS->enterFile(file);
+                fref.data(*initialGS).strictLevel = pipeline::decideStrictLevel(*initialGS, fref, config->opts);
+                if (fref.id() >= globalStateHashes.size()) {
+                    globalStateHashes.resize(fref.id() + 1);
+                }
+            }
+            evictedHashes[fref.id()] = move(globalStateHashes[fref.id()]);
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: This is where we save file hashes that this edit evicts.

File: main/lsp/request_dispatch.cc [hunk 23]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+        }
+    }
+
+    ENFORCE(update.updatedFiles.size() == update.updatedFileHashes.size());
+    ENFORCE(update.updatedFiles.size() == update.updatedFileIndexes.size());
+
+    // Completely replace `pendingTypecheckUpdates` if this was a slow path update.
+    if (!update.canTakeFastPath) {
+        update.updatedGS = initialGS->deepCopy();
+        pendingTypecheckUpdates = update.copy();
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: `pendingTypecheckUpdates` contains all of the edits that were last run on the slow path.

If the slow path finishes, it may be stale -- but that's OK. We only ever use it if the `epochManager` says we are running a slow path, and we succeed in canceling it.

File: main/lsp/request_dispatch.cc [hunk 24]
```diff
@@ -64,11 +321,22 @@ void LSPLoop::processRequestInternal(LSPMessage &msg) {
         auto &params = msg.asNotification().params;
         if (method == LSPMethod::SorbetWorkspaceEdit) {
             auto &editParams = get<unique_ptr<SorbetWorkspaceEditParams>>(params);
-            typecheckerCoord.typecheckOnSlowPath(move(editParams->updates));
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: (You may be concerned that previously we did not have a `FastPathTypecheckTask`. I had that thought initially, too. Then I noticed that the previous implementation of `typecheckOnSlowPath` just calls `LSPTypechecker::typecheck`, which takes the fast path if it's a fast path update.)

File: main/lsp/request_dispatch.cc [hunk 25]
```diff
@@ -64,11 +321,22 @@ void LSPLoop::processRequestInternal(LSPMessage &msg) {
...
+                // The slow path is non-blocking.
+                typecheckerCoord.typecheckOnSlowPath(move(updates));
+            }
         } else if (method == LSPMethod::Initialized) {
             prodCategoryCounterInc("lsp.messages.processed", "initialized");
-            auto &initParams = get<unique_ptr<InitializedParams>>(params);
-            typecheckerCoord.initialize(move(initParams));
+            LSPFileUpdates updates;
+            // Note: Captures all arguments by reference, and updates `initialGS`, `globalStateHashes`, and `updates`.
+            typecheckerCoord.syncRun(
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: This is the blocking part of initialization.

File: main/lsp/request_dispatch.cc [hunk 26]
```diff
@@ -64,11 +321,22 @@ void LSPLoop::processRequestInternal(LSPMessage &msg) {
...
+                make_unique<IndexAndHashWorkspaceTask>(*config, initialGS, globalStateHashes, updates));
+            // Send processed updates to typechecker (this will eventually be non-blocking w/ preemption).
+            typecheckerCoord.initialize(move(updates));
```
Comment (jvilk-stripe) [inline main/lsp/request_dispatch.cc]: This will be the non-blocking portion.

File: main/lsp/test/generate_lsp_messages_test.cc [hunk 27]
```diff
@@ -304,4 +305,55 @@ TEST(GenerateLSPMessagesTest, AcceptsNullOnOptionalFields) {
     parseTest<ConfigurationItem>("{\"scopeUri\": null}", [](auto &item) -> void { ASSERT_FALSE(item->scopeUri); });
 }
 
+TEST(GenerateLSPMessagesTest, SorbetWorkspaceEditParamsMerge) {
```
Comment (jvilk-stripe) [inline main/lsp/test/generate_lsp_messages_test.cc]: Tests for the new `SorbetWorkspaceEditParams::merge` method!

File: main/lsp/test/lsp_preprocessor_test.cc [hunk 28]
```diff
@@ -125,116 +106,6 @@ class CountingTask final : public core::lsp::Task {
 
 } // namespace
 
-TEST(TimeTravelingGlobalState, ComesBefore) { // NOLINT
```
Comment (jvilk-stripe) [inline main/lsp/test/lsp_preprocessor_test.cc]: I deleted all of the tests involving `TimeTravelingGlobalState` and fast path decisions, since the preprocessor is no longer involved with those matters.

File: main/lsp/tools/make_lsp_types.cc [hunk 29]
```diff
@@ -1244,11 +1244,7 @@ void makeLSPTypes(vector<shared_ptr<JSONClassType>> &enumTypes, vector<shared_pt
                    classTypes);
 
     // Empty object.
-    auto InitializedParams = makeObject("InitializedParams", {}, classTypes,
```
Comment (jvilk-stripe) [inline main/lsp/tools/make_lsp_types.cc]: The preprocessor used to populate these parameters with `LSPFileUpdates`. Not anymore!

File: main/lsp/request_dispatch.cc (hunk 22 above)
Comment (mkillianey-stripe) [reply main/lsp/request_dispatch.cc]: Since `updatedFiles` is a vector, not a set, checking...is there a guarantee that its values are unique?  (Or more importantly, that the fref is unique?)

If not, it looks like a file that was repeated (or resolved to a repeated fref) would evict the `evictedHashes` value from the first time around...is that possible, and if so, is that problematic?

File: main/lsp/request_dispatch.cc [hunk 30]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+    }
+
+    // Index changes in initialGS. pipeline::index sorts output by file id, but we need to reorder to match the order of
+    // other fields.
+    UnorderedMap<u2, int> fileToPos;
+    {
+        int i = -1;
+        for (auto fref : frefs) {
+            // We should have ensured before reaching here that there are no duplicates.
+            ENFORCE(!fileToPos.contains(fref.id()));
```
Comment (mkillianey-stripe) [inline main/lsp/request_dispatch.cc]: Ah, I guess this answers my question from above.  :-)
Does this ENFORCE kill just this thread (in which case, do we need to worry about mutations above?), or does it kill the running Sorbet entirely (in which case, I suppose it doesn't matter where it goes?)

File: main/lsp/LSPPreprocessor.h [hunk 31]
```diff
@@ -69,27 +51,27 @@ class LSPPreprocessor final {
...
+                                                                 std::unique_ptr<DidOpenTextDocumentParams> openParams);
+    std::unique_ptr<SorbetWorkspaceEditParams>
+    canonicalizeEdits(u4 v, std::unique_ptr<DidCloseTextDocumentParams> closeParams);
+    std::unique_ptr<SorbetWorkspaceEditParams>
+    canonicalizeEdits(u4 v, std::unique_ptr<WatchmanQueryResponse> queryResponse) const;
 
     /**
-     * Returns a global state for typechecking, cloned from initialGS. Note: The clone does not share an error queue
-     * with initialGS.
+     * Get the current contents of the file at the given path. Returns "" (empty string view) if file does not yet
```
[the remark this answers is not in the document]
Comment (jvilk-stripe) [reply main/lsp/LSPPreprocessor.h]: Good question. The `require` statement question isn't something that is handled by Sorbet; you hand Sorbet a set of files in the workspace and it builds up the global namespace.

What this method deals with is file URIs received from the client in `textDocument/didChange` events. By the standard, a `didChange` event should always be preceded by a `didOpen` event that contains the full text of the file. The reason being is that `didChange` supports diffs from the last-known state of the file; we can't know the last known state of the file if we didn't at some point receive the whole file. And we can't go to the disk for the whole file contents, as it might be a new file that hasn't been saved to disk.

So what you're seeing here is the following: The code right now treats a `didChange` to a file that wasn't previously opened as if the file contains nothing. This is the same behavior that we had prior to this change.

I think the _correct_ thing to do would be to send back an error somehow.

File: main/lsp/LSPTypechecker.cc (hunk 8 above)
Comment (jvilk-stripe) [reply main/lsp/LSPTypechecker.cc]: That's it.

File: main/lsp/LSPTypecheckerCoordinator.cc (hunk 13 above)
Comment (jvilk-stripe) [reply main/lsp/LSPTypecheckerCoordinator.cc]: >  Or by "now," do you mean, "it currently doesn't make sense, but it will in the future"?

It currently doesn't make sense. It will make sense in the future.

> In that case, don't you still want the old comment, because there will always be a synchronous part?

Nope! `InitializeTask` itself will be asynchronous. I was mistaken when we chatted before; `InitializeTask` is basically a specialized run slow path. The synchronous portion is the indexing and hashing task in request_dispatch, which previously ran in the preprocessor.

File: main/lsp/json_types.h [hunk 32]
```diff
@@ -18,15 +18,21 @@ namespace sorbet::realmain::lsp {
  */
 class LSPFileUpdates final {
 public:
-    // This specific update contains versions [versionStart, versionEnd].
-    u4 versionEnd = 0;
-    u4 versionStart = 0;
+    // This specific update contains edits with the given epoch
+    u4 epoch = 0;
+    // The total number of edits that this update represents. Used for stats.
+    u2 editCount = 0;
```
[the remark this answers is not in the document]
Comment (jvilk-stripe) [reply main/lsp/json_types.h]: Good point -- in practice this value likely will be passed out to use 4 bytes. This number is used for stats only, so I figured I'd use the smallest unsigned integer value that would handle 99.99% of cases. 64K could potentially happen during git operation (edits can be watchman events, too).

I'm fine with padding it out to u4.

File: main/lsp/lsp.h (hunk 17 above)
Comment (jvilk-stripe) [reply main/lsp/lsp.h]: I had an idea that maybe it would make sense to have an `LSPIndexer` that is similar to `LSPTypechecker` that owns all of this nonsense. Maybe in a future PR.

File: main/lsp/request_dispatch.cc [hunk 33]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+            // Clear error queue.
+            // (Note: Flushing is disabled in LSP mode, so we have to drain.)
+            initialGS->errorQueue->drainWithQueryResponses();
+        }
+
+        // When inputFileNames is 0 (as in tests), indexed ends up being size 0 because we don't index payload files.
+        // At the same time, we expect indexed to be the same size as GlobalStateHash, which _does_ have payload files.
+        // Resize the indexed array accordingly.
+        if (indexed.size() < initialGS->getFiles().size()) {
+            indexed.resize(initialGS->getFiles().size());
```
[the remark this answers is not in the document]
Comment (jvilk-stripe) [reply main/lsp/request_dispatch.cc]: Hah, good point. We are being a bit boneheaded here. I can address this in a subsequent PR.

File: main/lsp/request_dispatch.cc [hunk 34]
```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
...
+UnorderedMap<int, core::FileHash> mergeEvictions(const UnorderedMap<int, core::FileHash> &olderEvictions,
+                                                 const UnorderedMap<int, core::FileHash> &newerEvictions) {
+    // For evictions, which are needed for emulating an older `globalStateHashes`, we keep the oldest.
+    UnorderedMap<int, core::FileHash> combinedEvictions = olderEvictions;
+    for (auto &e : newerEvictions) {
+        if (!combinedEvictions.contains(e.first)) {
+            combinedEvictions[e.first] = e.second;
+        }
+    }
+    return combinedEvictions;
```
[the remark this answers is not in the document]
Comment (jvilk-stripe) [reply main/lsp/request_dispatch.cc]: Good call; I can use that.

File: main/lsp/request_dispatch.cc (hunk 22 above)
Comment (jvilk-stripe) [reply main/lsp/request_dispatch.cc]: Good question. `updatedFiles` should not contain any duplicated files. It's a vector for efficiency reasons, but is effectively treated like a set.

`SorbetWorkspaceEditParams` are created from four sources:

* didOpen messages, which concern one file.
* didChange messages, which concern one file.
* didClose messages, which concern one file.
* Watchman messages, which concern a set of files.

Each of these are converted into `SorbetWorkspaceEditParams` by `LSPPreprocessor`, and are merged with adjacent `SorbetWorkspaceEditParams` via `SorbetWorkspaceEditParams::merge`. That helper function ensures that `SorbetWorkspaceEditParams.updates` contains no duplicates during merging.

So, tl;dr: It shouldn't be possible. But it may be worth adding a few sanity checks here and there.

File: main/lsp/request_dispatch.cc (hunk 30 above)
Comment (jvilk-stripe) [reply main/lsp/request_dispatch.cc]: It kills the thread, but only in debug/test builds. In release builds, all `ENFORCE` statements are stripped out -- so we can liberally sprinkle them around.

They are meant to check invariants that _should always be true_ in your code.

Comment (jvilk-stripe) [reply main/lsp/request_dispatch.cc]: (ENFORCE also has some smarts regarding the debugger. If it detects an attached debugger on a failed ENFORCE, it will pause execution for inspection.)

File: main/lsp/request_dispatch.cc (hunk 34 above)
[the remark this answers is not in the document]
Comment (jvilk-stripe) [reply main/lsp/request_dispatch.cc]: On second thought, I'm not sure if we can use that.

We're using `absl::UnorderedMap`, and I don't see docs that describe `insert` behavior (and the source code is sufficiently abstract to be hard to follow). To be safe, I'll keep the current, more verbose version.
